### PR TITLE
Add Docker development environment

### LINKED ISSUE #57: Docker development environment
Replace our current Vagrant/Puppet development environment with a Docker one.

**Benefits**

1. With [Docker for Mac](https://blog.docker.com/2016/03/docker-for-mac-windows-beta/) around the corner, we can ditch VirtualBox. Docker for Mac uses the xhyve hypervisor, which is much faster and memory-efficient than VirtualBox. Moreover, it’s very easy to install and start.

2. Third-party developers will get ready-to-go images instead of being dependent on the Driebit Puppet master for provisioning.

3. Xhyve brings fast file sharing and supports inotify, so we get automatic recompilation (#27).

4. We can build on the Dockerfile from the Zotonic project instead of having a configuration from scratch.

**Proposal**

* Create a lightweight Zotonic image, without PostgreSQL: zotonic/zotonic#1295.
* Add a multi-container `docker-compose.yml` to Ginger. For now: one Dockerfile for all Ginger sites together.
* Perhaps later, run each site in its own container.

### PR DESCRIPTION
**Pull request by ddeboer**
* Fix #57.
* Upstream work underway to create a lightweight Zotonic image:  https://github.com/zotonic/zotonic/issues/1295.
* Driebit-namespaced Zotonic image (for Ginger-specific things) is in a [separate repository](https://github.com/driebit/docker-zotonic).

**Todo:**

- [x] separate NPM/Gulp/Bower container for running gulp: See [driebit/node-gulp](https://github.com/driebit/docker-node-gulp).
- [x] find a way for core Zotonic development: perhaps mount a host volume over `/opt/zotonic`? We probably want to get new deps and compile when users start switching between Zotonic branches
- [x] test [database import scripts](https://github.com/driebit/ginger/tree/master/scripts): how to import a database (requires `psql` binary) while having access to downloaded/copied SQL dumps? Add a volume to the postgres container?

**Issues:**

* inotify is having problems with the huge node_modules directories in our sites. Unfortunately, we cannot *exclude* them (`--exclude` in inotify seems to exclude file change notifications, but still watches the files and so counts against `max_user_watches`)

### CODE REVIEW

**Comment by ddeboer** [comment]
Okay, this is now functional:

- a development environment with inotify enabled (so no more flushing and `z:m()`0ing)
- seems to work okay with multiple websites enabled
- a separate Docker image for running Gulp.

Anyone up to testing this before we merge to master? Docker Beta is still closed, so you’ll need a [Docker Beta token](https://beta.docker.com/) for now.

**Comment by ddeboer** [comment]
Local Zotonic development now possible!💥

**Comment by Dirklectisch** [comment]
I signed up for Docket Beta yesterday, will test as soon as I get in.

**Comment by Dirklectisch** [comment]
Tested this out on two computers and it seems to work. I did have it crash once but wasn't able to reproduce the problem after that. It logged the following error:

````
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 439, in wait
  File "site-packages/docker/client.py", line 154, in _raise_for_status
APIError: 502 Server Error: Bad Gateway ("Bad response from Docker engine")
Exception in thread Thread-3:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 439, in wait
  File "site-packages/docker/client.py", line 154, in _raise_for_status
APIError: 502 Server Error: Bad Gateway ("Bad response from Docker engine")
````

**Comment by ddeboer** [comment]
> It logged the following error:

Looks like a Docker thing: ‘Bad response from Docker engine’. Just try to upgrade Docker – seems they have a new version every day now. :wink:

**Comment by ddeboer** [comment]
As @Dirklectisch correctly noted, the database import scripts weren’t updated for the Docker config. I did so now: see the changes in the README. If you wish to try this out, you may have to throw away your old branch, as I squashed this one.

File: README.md [hunk 1]
```diff
@@ -4,89 +4,98 @@ Ginger
...
-```
+### Import a database from a remote backup
+
+1. If there are no backups yet, create a backup on the remote Zotonic site.
+
+2. Then run:
+
+    ```bash
+    $ make import-db-backup host=ginger01.driebit.net site=site-name
+    ```
```
Comment (ddeboer) [inline README.md]: Moved the import scripts to a `Makefile` for re-usability.

**Comment by ddeboer** [comment]
Docker for Mac [now in public beta](https://blog.docker.com/2016/06/docker-mac-windows-public-beta/), so have fun! 😄

**Pull request merged**